### PR TITLE
[SQL Lab] Adding DB options for SQL LAb

### PR DESCRIPTION
**Pull request by mistercrunch**
screen shot 2016-08-31 at 5 52 29 pm

each db can be exposed or not in SQL Lab
CTAS is an option
target_schema placeholder (not hooked yet, but would force the CTAS to
target a specific schema)

### CODE REVIEW

File: caravel/assets/javascripts/SqlLab/components/SqlEditor.jsx [hunk 1]
```diff
@@ -207,31 +207,37 @@ class SqlEditor extends React.Component {
...
+        <FormGroup>
+          <InputGroup>
+            <FormControl
+              type="text"
+              bsSize="small"
+              className="input-sm"
+              placeholder="new table name"
+              onChange={this.ctasChanged.bind(this)}
+            />
+            <InputGroup.Button>
```
Comment (bkyryliuk) [inline caravel/assets/javascripts/SqlLab/components/SqlEditor.jsx]: could you please explain why do you have   <InputGroup.Button> and <Button> ?

File: caravel/assets/javascripts/SqlLab/components/SqlEditor.jsx [hunk 2]
```diff
@@ -276,9 +282,10 @@ class SqlEditor extends React.Component {
 }
 
 SqlEditor.propTypes = {
-  queryEditor: React.PropTypes.object,
   actions: React.PropTypes.object,
+  database: React.PropTypes.object,
```
Comment (bkyryliuk) [inline caravel/assets/javascripts/SqlLab/components/SqlEditor.jsx]: could you please add a comment specifying the schema of the object ?

File: caravel/assets/javascripts/SqlLab/components/SqlEditorLeft.jsx [hunk 3]
```diff
@@ -73,11 +77,17 @@ class SqlEditorTopToolbar extends React.Component {
   }
   fetchDatabaseOptions() {
     this.setState({ databaseLoading: true });
-    const url = '/databaseasync/api/read';
+    const url = '/databaseasync/api/read?_flt_0_expose_in_sqllab=1';
```
Comment (bkyryliuk) [inline caravel/assets/javascripts/SqlLab/components/SqlEditorLeft.jsx]: I didn't find the _flt_0_expose_in_sqllab arg, could you please explain it in the comment ?

File: caravel/migrations/versions/41f6a59a61f2_database_options_for_sql_lab.py [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+down_revision = '3c3ffe173e4f'
+
+
+def upgrade():
+    op.add_column('dbs', sa.Column('allow_ctas', sa.Boolean(), nullable=True))
+    op.add_column(
+        'dbs', sa.Column('expose_in_sqllab', sa.Boolean(), nullable=True))
+    op.add_column(
+        'dbs',
+        sa.Column('force_ctas_schema', sa.String(length=250), nullable=True))
```
Comment (bkyryliuk) [inline caravel/migrations/versions/41f6a59a61f2_database_options_for_sql_lab.py]: s/force_ctas_schema/ctas_schema
I would expect to have a var that starts with force_ to be boolean. Optional.

File: caravel/migrations/versions/41f6a59a61f2_database_options_for_sql_lab.py [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+
+
+def downgrade():
+    op.drop_column('dbs', 'force_ctas_schema')
```
Comment (bkyryliuk) [inline caravel/migrations/versions/41f6a59a61f2_database_options_for_sql_lab.py]: this operation would most likely fail in sqllite.
eventually we may want to add a downgrade test.

File: caravel/models.py [hunk 6]
```diff
@@ -1574,9 +1577,14 @@ def wrapper(*args, **kwargs):
             d.update(kwargs)
             slice_id = d.get('slice_id', 0)
             slice_id = int(slice_id) if slice_id else 0
+            params = ""
+            try:
+                params = json.dumps(d)
```
Comment (bkyryliuk) [inline caravel/models.py]: what about trying to apply str(arg) to every arg rather than catching the exception here?

File: caravel/sql_lab.py [hunk 7]
```diff
@@ -100,10 +100,11 @@ def get_sql_results(query_id, return_results=True):
...
         data = None
         if result_proxy.cursor:
-            cols = [col[0] for col in result_proxy.cursor.description]
+            columns = [col[0] for col in result_proxy.cursor.description]
             data = result_proxy.fetchall()
-            df = pd.DataFrame(data, columns=cols)
-            columns = [c for c in df.columns]
+            df = pd.DataFrame(data, columns=columns)
+            df = df.where((pd.notnull(df)), None)
+            # data = df.values.tolist()
```
Comment (bkyryliuk) [inline caravel/sql_lab.py]: delete this line

File: caravel/sql_lab.py [hunk 8]
```diff
@@ -100,10 +100,11 @@ def get_sql_results(query_id, return_results=True):
         columns = None
         data = None
         if result_proxy.cursor:
-            cols = [col[0] for col in result_proxy.cursor.description]
+            columns = [col[0] for col in result_proxy.cursor.description]
             data = result_proxy.fetchall()
-            df = pd.DataFrame(data, columns=cols)
-            columns = [c for c in df.columns]
+            df = pd.DataFrame(data, columns=columns)
+            df = df.where((pd.notnull(df)), None)
```
Comment (bkyryliuk) [inline caravel/sql_lab.py]: s/df/ndf or any other name different from the df

File: caravel/utils.py [hunk 9]
```diff
@@ -90,6 +91,22 @@ def __get__(self, obj, objtype):
         return functools.partial(self.__call__, obj)
 
 
+def get_or_create_db(caravel):
+    db = caravel.db
+    config = caravel.app.config
+    DB = caravel.models.Database
+    print("Creating database reference")
```
Comment (bkyryliuk) [inline caravel/utils.py]: log instead of printing

File: caravel/utils.py [hunk 10]
```diff
@@ -90,6 +91,22 @@ def __get__(self, obj, objtype):
...
+    dbobj = db.session.query(DB).filter_by(database_name='main').first()
+    if not dbobj:
+        dbobj = DB(database_name="main")
+    print(config.get("SQLALCHEMY_DATABASE_URI"))
```
Comment (bkyryliuk) [inline caravel/utils.py]: ditto

File: caravel/utils.py [hunk 11]
```diff
@@ -90,6 +91,22 @@ def __get__(self, obj, objtype):
         return functools.partial(self.__call__, obj)
 
 
+def get_or_create_db(caravel):
```
Comment (bkyryliuk) [inline caravel/utils.py]: maybe s/get_or_create_db/get_or_create_main_db

File: caravel/assets/javascripts/SqlLab/components/SqlEditorLeft.jsx (hunk 3 above)
Comment (mistercrunch) [inline caravel/assets/javascripts/SqlLab/components/SqlEditorLeft.jsx]: it's part of the FAB rest mini language, not sure if we need to explain it in this context (and all other contexes where it's used)

File: caravel/migrations/versions/41f6a59a61f2_database_options_for_sql_lab.py (hunk 4 above)
Comment (mistercrunch) [inline caravel/migrations/versions/41f6a59a61f2_database_options_for_sql_lab.py]: I think it's fair to assume that defining this field flips the switch and that leaving it empty means it's not enforced. I'll make it clearer in the field description.

File: caravel/models.py (hunk 6 above)
Comment (mistercrunch) [inline caravel/models.py]: yes I don't think this is needed anymore, I did this so I could move forward, I think I can turn it off now.

File: caravel/sql_lab.py (hunk 8 above)
Comment (mistercrunch) [inline caravel/sql_lab.py]: I think it's fair to mutate it here as I don't need the old dataframe anymore, it's a pretty common pattern in pandas to keep re-affecting df.

File: caravel/assets/javascripts/SqlLab/components/SqlEditor.jsx (hunk 2 above)
Comment (mistercrunch) [inline caravel/assets/javascripts/SqlLab/components/SqlEditor.jsx]: let's define where we store schemas (maybe as functions that return objects)

File: caravel/migrations/versions/41f6a59a61f2_database_options_for_sql_lab.py (hunk 5 above)
Comment (mistercrunch) [inline caravel/migrations/versions/41f6a59a61f2_database_options_for_sql_lab.py]: Let's do all of them in another PR

**Comment by mistercrunch** [comment]
Addressed all comments, mergin'!
